src/WP_CLI/RecursiveDataStructureTraverser.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+namespace WP_CLI;
+
+class RecursiveDataStructureTraverser {
+
+	/**
+	 * @var mixed The data to traverse set by reference.
+	 */
+	protected $data;
+
+	/**
+	 * @var string The character/sequence used to delineate hierarchy in a single key.
+	 */
+	protected $delimiter;
+
+	/**
+	 * @var null|string The key the data belongs to in the parent's data.
+	 */
+	protected $key;
+
+	/**
+	 * @var null|static The parent instance of the traverser.
+	 */
+	protected $parent;
+
+	/**
+	 * RecursiveDataStructureTraverser constructor.
+	 *
+	 * @param        $data
+	 * @param static $parent
+	 */
+	public function __construct( &$data, $parent = null ) {
+		$this->data =& $data;
+		$this->parent = $parent;
+	}
+
+	/**
+	 * @param $locator
+	 *
+	 * @throws \Exception
+	 *
+	 * @return static
+	 */
+	public function get( $locator ) {
+		return $this->traverse_to( (array) $locator )->value();
+	}
+
+	/**
+	 * Get the current data.
+	 *
+	 * @return mixed
+	 */
+	public function value() {
+		return $this->data;
+	}
+
+	public function update( $locator, $value ) {
+		$this->traverse_to( (array) $locator )->set_value( $value );
+	}
+
+	public function set_value( $value ) {
+		$this->data = $value;
+	}
+
+	public function delete( $locator ) {
+		$this->traverse_to( (array) $locator )->unset_on_parent();
+	}
+
+	public function insert( $locator, $value ) {
+		try {
+			$this->update( $locator, $value );
```

Comment:
I wouldn't say it falls back to `patch update`, but it is the same operation with the exception that it creates any missing keys along the way. If the missing key exception is caught, it creates it and calls itself again.